File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/Code/ExpressionType.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#pragma once

#include "ExpressionType.h"




// Math
const std::string ExpressionType::Add = "+";
const std::string ExpressionType::Subtract = "-";
const std::string ExpressionType::Multiply = "*";
const std::string ExpressionType::Divide = "/";
const std::string ExpressionType::Min = "min";
const std::string ExpressionType::Max = "max";
const std::string ExpressionType::Power = "^";
const std::string ExpressionType::Mod = "%";
const std::string ExpressionType::Average = "average";
const std::string ExpressionType::Sum = "sum";
const std::string ExpressionType::Range = "range";
const std::string ExpressionType::Floor = "floor";
const std::string ExpressionType::Ceiling = "ceiling";
const std::string ExpressionType::Round = "round";

// Comparisons
const std::string ExpressionType::LessThan = "<";
const std::string ExpressionType::LessThanOrEqual = "<=";
const std::string ExpressionType::Equal = "==";
const std::string ExpressionType::NotEqual = "!=";
const std::string ExpressionType::GreaterThan = ">";
const std::string ExpressionType::GreaterThanOrEqual = ">=";
const std::string ExpressionType::Exists = "exists";

// Logic
const std::string ExpressionType::And = "&&";
const std::string ExpressionType::Or = "||";
const std::string ExpressionType::Not = "!";

// String
const std::string ExpressionType::Concat = "concat";
const std::string ExpressionType::Length = "length";
const std::string ExpressionType::Replace = "replace";
const std::string ExpressionType::ReplaceIgnoreCase = "replaceIgnoreCase";
const std::string ExpressionType::Split = "split";
const std::string ExpressionType::Substring = "substring";
const std::string ExpressionType::ToLower = "toLower";
const std::string ExpressionType::ToUpper = "toUpper";
const std::string ExpressionType::Trim = "trim";
const std::string ExpressionType::EndsWith = "endsWith";
const std::string ExpressionType::StartsWith = "startsWith";
const std::string ExpressionType::CountWord = "countWord";
const std::string ExpressionType::AddOrdinal = "addOrdinal";
const std::string ExpressionType::NewGuid = "newGuid";
const std::string ExpressionType::IndexOf = "indexOf";
const std::string ExpressionType::LastIndexOf = "lastIndexOf";
const std::string ExpressionType::EOL = "EOL";
const std::string ExpressionType::SentenceCase = "sentenceCase";
const std::string ExpressionType::TitleCase = "titleCase";

// Collection
const std::string ExpressionType::Count = "count";
const std::string ExpressionType::Contains = "contains";
const std::string ExpressionType::Empty = "empty";
const std::string ExpressionType::Join = "join";
const std::string ExpressionType::First = "first";
const std::string ExpressionType::Last = "last";
const std::string ExpressionType::Foreach = "foreach";
const std::string ExpressionType::Select = "select";
const std::string ExpressionType::Where = "where";
const std::string ExpressionType::Union = "union";
const std::string ExpressionType::Intersection = "intersection";
const std::string ExpressionType::Skip = "skip";
const std::string ExpressionType::Take = "take";
const std::string ExpressionType::SubArray = "subArray";
const std::string ExpressionType::SortBy = "sortBy";
const std::string ExpressionType::SortByDescending = "sortByDescending";
const std::string ExpressionType::IndicesAndValues = "indicesAndValues";
const std::string ExpressionType::Flatten = "flatten";
const std::string ExpressionType::Unique = "unique";

// DateTime
const std::string ExpressionType::AddDays = "addDays";
const std::string ExpressionType::AddHours = "addHours";
const std::string ExpressionType::AddMinutes = "addMinutes";
const std::string ExpressionType::AddSeconds = "addSeconds";
const std::string ExpressionType::DayOfMonth = "dayOfMonth";
const std::string ExpressionType::DayOfWeek = "dayOfWeek";
const std::string ExpressionType::DayOfYear = "dayOfYear";
const std::string ExpressionType::Month = "month";
const std::string ExpressionType::Date = "date";
const std::string ExpressionType::Year = "year";
const std::string ExpressionType::UtcNow = "utcNow";
const std::string ExpressionType::FormatDateTime = "formatDateTime";
const std::string ExpressionType::FormatEpoch = "formatEpoch";
const std::string ExpressionType::FormatTicks = "formatTicks";
const std::string ExpressionType::SubtractFromTime = "subtractFromTime";
const std::string ExpressionType::DateReadBack = "dateReadBack";
const std::string ExpressionType::GetTimeOfDay = "getTimeOfDay";
const std::string ExpressionType::GetFutureTime = "getFutureTime";
const std::string ExpressionType::GetPastTime = "getPastTime";
const std::string ExpressionType::ConvertFromUtc = "convertFromUTC";
const std::string ExpressionType::ConvertToUtc = "convertToUTC";
const std::string ExpressionType::AddToTime = "addToTime";
const std::string ExpressionType::StartOfDay = "startOfDay";
const std::string ExpressionType::StartOfHour = "startOfHour";
const std::string ExpressionType::StartOfMonth = "startOfMonth";
const std::string ExpressionType::Ticks = "ticks";
const std::string ExpressionType::TicksToDays = "ticksToDays";
const std::string ExpressionType::TicksToHours = "ticksToHours";
const std::string ExpressionType::TicksToMinutes = "ticksToMinutes";
const std::string ExpressionType::DateTimeDiff = "dateTimeDiff";

// Timex 
const std::string ExpressionType::IsDefinite = "isDefinite";
const std::string ExpressionType::IsTime = "isTime";
const std::string ExpressionType::IsDuration = "isDuration";
const std::string ExpressionType::IsDate = "isDate";
const std::string ExpressionType::IsTimeRange = "isTimeRange";
const std::string ExpressionType::IsDateRange = "isDateRange";
const std::string ExpressionType::IsPresent = "isPresent";
const std::string ExpressionType::GetNextViableDate = "getNextViableDate";
const std::string ExpressionType::GetPreviousViableDate = "getPreviousViableDate";
const std::string ExpressionType::GetNextViableTime = "getNextViableTime";
const std::string ExpressionType::GetPreviousViableTime = "getPreviousViableTime";

// Conversions
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
const std::string ExpressionType::Float = "float";
#pragma warning restore CA1720 // Identifier contains type name
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
const std::string ExpressionType::Int = "int";
#pragma warning restore CA1720 // Identifier contains type name
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
const std::string ExpressionType::String = "string";
#pragma warning restore CA1720 // Identifier contains type name
const std::string ExpressionType::JsonStringify = "jsonStringify";
const std::string ExpressionType::Bool = "bool";
const std::string ExpressionType::Binary = "binary";
const std::string ExpressionType::Base64 = "base64";
const std::string ExpressionType::Base64ToBinary = "base64ToBinary";
const std::string ExpressionType::Base64ToString = "base64ToString";
const std::string ExpressionType::DataUri = "dataUri";
const std::string ExpressionType::DataUriToBinary = "dataUriToBinary";
const std::string ExpressionType::DataUriToString = "dataUriToString";
const std::string ExpressionType::UriComponent = "uriComponent";
const std::string ExpressionType::UriComponentToString = "uriComponentToString";
const std::string ExpressionType::Xml = "xml";
const std::string ExpressionType::FormatNumber = "formatNumber";

// URI Parsing Functions
const std::string ExpressionType::UriHost = "uriHost";
const std::string ExpressionType::UriPath = "uriPath";
const std::string ExpressionType::UriPathAndQuery = "uriPathAndQuery";
const std::string ExpressionType::UriPort = "uriPort";
const std::string ExpressionType::UriQuery = "uriQuery";
const std::string ExpressionType::UriScheme = "uriScheme";

// Memory
const std::string ExpressionType::Accessor = "Accessor";
const std::string ExpressionType::Element = "Element";
const std::string ExpressionType::CreateArray = "createArray";

// Misc
const std::string ExpressionType::Constant = "Constant";
const std::string ExpressionType::Lambda = "Lambda";
const std::string ExpressionType::If = "if";
const std::string ExpressionType::Rand = "rand";

// Object manipulation and construction functions
const std::string ExpressionType::Json = "json";
const std::string ExpressionType::GetProperty = "getProperty";
const std::string ExpressionType::AddProperty = "addProperty";
const std::string ExpressionType::RemoveProperty = "removeProperty";
const std::string ExpressionType::SetProperty = "setProperty";
const std::string ExpressionType::Coalesce = "coalesce";
const std::string ExpressionType::XPath = "xPath";
const std::string ExpressionType::SetPathToValue = "setPathToValue";
const std::string ExpressionType::JPath = "jPath";
const std::string ExpressionType::Merge = "merge";

// Regular expression
const std::string ExpressionType::IsMatch = "isMatch";

//Type Checking
const std::string ExpressionType::IsInteger = "isInteger";
const std::string ExpressionType::IsFloat = "isFloat";
const std::string ExpressionType::IsString = "isString";
const std::string ExpressionType::IsArray = "isArray";
const std::string ExpressionType::IsObject = "isObject";
const std::string ExpressionType::IsBoolean = "isBoolean";
const std::string ExpressionType::IsDateTime = "isDateTime";

// trigger tree 

/// <summary>
/// Mark a sub-expression as optional.
/// </summary>
/// <remarks>
/// When an expression is being processed, optional creates a disjunction where the expression is both included and not
/// included with the rest of the expression.  This is a simple way to express this common relationship.  By generating
/// both clauses then matching the expression can be more specific when the optional expression is true.
/// </remarks>
const std::string ExpressionType::Optional = "optional";

/// <summary>
/// Any predicate expression wrapped in this will be ignored for specialization.
/// </summary>
/// <remarks>
/// This is useful for when you need to add expression to the trigger that are part of rule mechanics rather than of intent.
/// For example, if you have a counter for how often a particular message is displayed, then that is part of the triggering condition, 
/// but all such rules would be incomparable because they counter is per-rule. 
/// </remarks>
const std::string ExpressionType::Ignore = "ignore";
#pragma warning restore CS1591 // Missing XML comment for publicly visible type or member

File: libraries/Microsoft.AdaptiveExpressions.Core/ExpressionType.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.AdaptiveExpressions.Core
{
    /// <summary>
    /// Built-in expression types.
    /// </summary>
    public static class ExpressionType
    {
        // Math
#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
        public const string Add = "+";
        public const string Subtract = "-";
        public const string Multiply = "*";
        public const string Divide = "/";
        public const string Min = "min";
        public const string Max = "max";
        public const string Power = "^";
        public const string Mod = "%";
        public const string Average = "average";
        public const string Sum = "sum";
        public const string Range = "range";
        public const string Floor = "floor";
        public const string Ceiling = "ceiling";
        public const string Round = "round";
        public const string Abs = "abs";
        public const string Sqrt = "sqrt";

        // Comparisons
        public const string LessThan = "<";
        public const string LessThanOrEqual = "<=";
        public const string Equal = "==";
        public const string NotEqual = "!=";
        public const string GreaterThan = ">";
        public const string GreaterThanOrEqual = ">=";
        public const string Exists = "exists";

        // Logic
        public const string And = "&&";
        public const string Or = "||";
        public const string Not = "!";

        // String
        public const string Concat = "concat";
        public const string Length = "length";
        public const string Replace = "replace";
        public const string ReplaceIgnoreCase = "replaceIgnoreCase";
        public const string Split = "split";
        public const string Substring = "substring";
        public const string ToLower = "toLower";
        public const string ToUpper = "toUpper";
        public const string Trim = "trim";
        public const string EndsWith = "endsWith";
        public const string StartsWith = "startsWith";
        public const string CountWord = "countWord";
        public const string AddOrdinal = "addOrdinal";
        public const string NewGuid = "newGuid";
        public const string IndexOf = "indexOf";
        public const string LastIndexOf = "lastIndexOf";
        public const string EOL = "EOL";
        public const string SentenceCase = "sentenceCase";
        public const string TitleCase = "titleCase";

        // Collection
        public const string Count = "count";
        public const string Contains = "contains";
        public const string Empty = "empty";
        public const string Join = "join";
        public const string First = "first";
        public const string Last = "last";
        public const string Foreach = "foreach";
        public const string Select = "select";
        public const string Where = "where";
        public const string Union = "union";
        public const string Intersection = "intersection";
        public const string Skip = "skip";
        public const string Take = "take";
        public const string SubArray = "subArray";
        public const string SortBy = "sortBy";
        public const string SortByDescending = "sortByDescending";
        public const string IndicesAndValues = "indicesAndValues";
        public const string Flatten = "flatten";
        public const string Unique = "unique";
        public const string Reverse = "reverse";
        public const string Any = "any";
        public const string All = "all";

        // DateTime
        public const string AddDays = "addDays";
        public const string AddHours = "addHours";
        public const string AddMinutes = "addMinutes";
        public const string AddSeconds = "addSeconds";
        public const string DayOfMonth = "dayOfMonth";
        public const string DayOfWeek = "dayOfWeek";
        public const string DayOfYear = "dayOfYear";
        public const string Month = "month";
        public const string Date = "date";
        public const string Year = "year";
        public const string UtcNow = "utcNow";
        public const string FormatDateTime = "formatDateTime";
        public const string FormatEpoch = "formatEpoch";
        public const string FormatTicks = "formatTicks";
        public const string SubtractFromTime = "subtractFromTime";
        public const string DateReadBack = "dateReadBack";
        public const string GetTimeOfDay = "getTimeOfDay";
        public const string GetFutureTime = "getFutureTime";
        public const string GetPastTime = "getPastTime";
        public const string ConvertFromUtc = "convertFromUTC";
        public const string ConvertToUtc = "convertToUTC";
        public const string AddToTime = "addToTime";
        public const string StartOfDay = "startOfDay";
        public const string StartOfHour = "startOfHour";
        public const string StartOfMonth = "startOfMonth";
        public const string Ticks = "ticks";
        public const string TicksToDays = "ticksToDays";
        public const string TicksToHours = "ticksToHours";
        public const string TicksToMinutes = "ticksToMinutes";
        public const string DateTimeDiff = "dateTimeDiff";

        // Timex 
        public const string IsDefinite = "isDefinite";
        public const string IsTime = "isTime";
        public const string IsDuration = "isDuration";
        public const string IsDate = "isDate";
        public const string IsTimeRange = "isTimeRange";
        public const string IsDateRange = "isDateRange";
        public const string IsPresent = "isPresent";
        public const string GetNextViableDate = "getNextViableDate";
        public const string GetPreviousViableDate = "getPreviousViableDate";
        public const string GetNextViableTime = "getNextViableTime";
        public const string GetPreviousViableTime = "getPreviousViableTime";
        public const string TimexResolve = "resolve";

        // Conversions
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
        public const string Float = "float";
#pragma warning restore CA1720 // Identifier contains type name
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
        public const string Int = "int";
#pragma warning restore CA1720 // Identifier contains type name
#pragma warning disable CA1720 // Identifier contains type name (by design and can't change this because of backward compat)
        public const string String = "string";
#pragma warning restore CA1720 // Identifier contains type name
        public const string JsonStringify = "jsonStringify";
        public const string Bool = "bool";
        public const string Binary = "binary";
        public const string Base64 = "base64";
        public const string Base64ToBinary = "base64ToBinary";
        public const string Base64ToString = "base64ToString";
        public const string DataUri = "dataUri";
        public const string DataUriToBinary = "dataUriToBinary";
        public const string DataUriToString = "dataUriToString";
        public const string UriComponent = "uriComponent";
        public const string UriComponentToString = "uriComponentToString";
        public const string Xml = "xml";
        public const string FormatNumber = "formatNumber";

        // URI Parsing Functions
        public const string UriHost = "uriHost";
        public const string UriPath = "uriPath";
        public const string UriPathAndQuery = "uriPathAndQuery";
        public const string UriPort = "uriPort";
        public const string UriQuery = "uriQuery";
        public const string UriScheme = "uriScheme";

        // Memory
        public const string Accessor = "Accessor";
        public const string Element = "Element";
        public const string CreateArray = "createArray";

        // Misc
        public const string Constant = "Constant";
        public const string Lambda = "Lambda";
        public const string If = "if";
        public const string Rand = "rand";

        // Object manipulation and construction functions
        public const string Json = "json";
        public const string GetProperty = "getProperty";
        public const string AddProperty = "addProperty";
        public const string RemoveProperty = "removeProperty";
        public const string SetProperty = "setProperty";
        public const string Coalesce = "coalesce";
        public const string XPath = "xPath";
        public const string SetPathToValue = "setPathToValue";
        public const string JPath = "jPath";
        public const string Merge = "merge";

        // Regular expression
        public const string IsMatch = "isMatch";

        //Type Checking
        public const string IsInteger = "isInteger";
        public const string IsFloat = "isFloat";
        public const string IsString = "isString";
        public const string IsArray = "isArray";
        public const string IsObject = "isObject";
        public const string IsBoolean = "isBoolean";
        public const string IsDateTime = "isDateTime";

        // StringOrValue
        public const string StringOrValue = "stringOrValue";

        // trigger tree 

        /// <summary>
        /// Mark a sub-expression as optional.
        /// </summary>
        /// <remarks>
        /// When an expression is being processed, optional creates a disjunction where the expression is both included and not
        /// included with the rest of the expression.  This is a simple way to express this common relationship.  By generating
        /// both clauses then matching the expression can be more specific when the optional expression is true.
        /// </remarks>
        public const string Optional = "optional";

        /// <summary>
        /// Any predicate expression wrapped in this will be ignored for specialization.
        /// </summary>
        /// <remarks>
        /// This is useful for when you need to add expression to the trigger that are part of rule mechanics rather than of intent.
        /// For example, if you have a counter for how often a particular message is displayed, then that is part of the triggering condition, 
        /// but all such rules would be incomparable because they counter is per-rule. 
        /// </remarks>
        public const string Ignore = "ignore";
#pragma warning restore CS1591 // Missing XML comment for publicly visible type or member
    }
}
